### PR TITLE
feat(forge): doc

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Motivation

Implementation of `forge doc` described in #1675 

## Solution

The `forge doc` command parses the source contracts and generates markdown files under `<root>/docs`. These files can be served by running `forge doc --serve`.

See [README](https://github.com/foundry-rs/foundry/blob/5d769fab5333c29e6acbf90e6dc948c9b04928d8/doc/README.md) for more details on the architecture of the `doc` module

### CODE REVIEW

File: cli/src/cmd/forge/doc.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                        ))?;
+                let mut doc = SolidityDoc::new(comments);
+                source_unit.visit(&mut doc)?;
+
+                Ok((path.clone(), doc.parts))
+            })
+            .collect::<eyre::Result<Vec<_>>>()?;
+
+        let out_dir = self.root.as_ref().unwrap_or(&find_project_root_path()?).join("docs");
+        fs::create_dir_all(&out_dir.join("src"))?;
```
Comment (tynes) [inline cli/src/cmd/forge/doc.rs:67]: Should this read the value from `foundry.toml` instead of hardcoding src?

Comment (rkrasiuk) [reply cli/src/cmd/forge/doc.rs:67]: yup, as well as the title & description for the docs. i put it out as a POC in order to receive comments regarding the implementation & output format. based on the comments, i will change it be more rustdoc style like

**Review by mattsse — COMMENTED** [review]
+1 on making `doc` a crate, let's try to move as much of the logic there,

ideally, there should be some kind of Builder, or function that accepts a config and will handle it from there.

File: doc/src/format.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+use crate::output::DocOutput;
+
+pub trait DocFormat {
+    fn doc(&self) -> String;
+}
+
+impl<'a> DocFormat for DocOutput<'a> {
+    fn doc(&self) -> String {
+        match self {
+            Self::H1(val) => format!("# {val}"),
```
Comment (sambacha) [inline doc/src/format.rs:16]: is it possible to add pandoc attribute headers here too? 
[see pandoc docs about attr headers](https://pandoc.org/MANUAL.html#extension-header_attributes)

Also, idk if you may find this useful reference https://github.com/sambacha/go-natspec , would like to see forge doc as a separate binary if possible (if its a separate crate this should be the case correct?)

**Comment by sambacha** [comment]
> Looking really great! Some comments from testing:
> 
> * [ ]  `forge init` should add the `docs` folder to the default `.gitignore`
> * [ ]  The [primitive-dodoc](https://github.com/primitivefinance/primitive-dodoc#-usage) hardhat plugin puts `@dev` tags in italics as shown there, open to other ideas though.
> * [ ]  The default homepage is called README and is empty. Perhaps rename to `Home` or something more user (non-dev) friendly
> * [ ]  Pull in the root `README.md` by default if an existing `docs/src/README.md` file is not present?
> * [ ]  Long functions are always one line and require scrolling. Would be more readable if they kept the original source formatting (with any inline comments?), or just apply the project's `forge fmt` settings to it for now if that's easier
> * [ ]  In a function's table, wdyt of putting the variable name and type as code instead, e.g. `name` and `uint256` instead of "name" and "uint256". Similar for state var/function headers. (Just an idea, open to pushback here)
> * [ ]  In `book.toml`, I recommend linking to the mdbook docs config page so people know available options. Perhaps some common ones should be exposed in `foundry.toml`, so you don't need to manually change `book.toml` everytime? Alternatively, let people have `docs/book.template.toml` file that gets used as the base?
> * [ ]  Is there a better way to differentiate folders and files in the sidebar, such as moving the arrows to the left of the label instead of right-aligned? Visually it looks very flat
> * [ ]  When inheriting from a contract: Files in the project link to their doc page, can dependencies link to the repo/commit?
> * [ ]  Similarly, wdyt about a way to link to a contract on github from it's docs page? Docgen would get current git commit and generate repo links based on that
> * [ ]  Inheritdoc support

Italics **shouldnt** be used for `@dev`, in general for typesetting we use italics for:

> - Quantities and variables — italic
> Symbols for quantities are italic, as are symbols for functions in general, for example, f (x ):[^1]

[^1]: https://physics.nist.gov/cuu/pdf/typefaces.pdf

Other styling related information I have documented here if your curious: https://github.com/sambacha/dappspec/tree/master/solidity/src


Foundry.toml is already overloaded with configurations, would it be possible to provide a link to book.toml for those not wanting to make those adjustments all the time for documentation related changes? I have CI triggers that ignore *.md, /docs, etc, so having book.toml makes it easier to not trigger documentation-related changes for workflows unrelated to that.

**Comment by mds1** [comment]
> Italics shouldnt be used for @dev, in general for typesetting we use italics for:

This is a typesetting guide for scientific manuscripts, so doesn't seem directly applicable. Regardless, I don't feel strongly about if/how `dev` and `notice` are distinguished, and use the italics suggestion because there's precedent there. Open to other suggestions if you have any

> would it be possible to provide a link to book.toml for those not wanting to make those adjustments all the time for documentation related changes? 

Not sure I'm following, are you suggesting having the foundry.toml allow something like this:

```toml
[doc]
config = "https://github.com/mds1/my-book-config/book.toml"
```

**Comment by sambacha** [comment]
> > Italics shouldnt be used for @dev, in general for typesetting we use italics for:
> 
> 
> 
> This is a typesetting guide for scientific manuscripts, so doesn't seem directly applicable. Regardless, I don't feel strongly about if/how `dev` and `notice` are distinguished, and used the italics suggestion because there's precedent there. Open to other suggestions if you have any
> 
> 
> 
> > would it be possible to provide a link to book.toml for those not wanting to make those adjustments all the time for documentation related changes? 
> 
> 
> 
> Not sure I'm following, are you suggesting having the foundry.toml allow something like this:
> 
> 
> 
> ```toml
> 
> [doc]
> 
> config = "https://github.com/mds1/my-book-config/book.toml"
> 
> ```

Just leave it unstyled, the first thing people will want to do is remove any styling and apply their own. 


Yes, though was thinking just a local pathname /docs/book.toml 

Making it independent of foundry.toml allows for it to be built with mdbook separately

File: config/src/doc.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+//! Configuration specific to the `forge doc` command and the `forge_doc` package
+
+use std::path::PathBuf;
+
```
Comment (mattsse) [inline config/src/doc.rs:4]:
```suggestion
```

File: doc/src/builder.rs [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+use itertools::Itertools;
+use rayon::prelude::*;
+use std::{
+    cmp::Ordering,
+    collections::HashMap,
+    fs,
+    path::{Path, PathBuf},
+};
+use toml::value;
+
```
Comment (mattsse) [inline doc/src/builder.rs:13]:
```suggestion
```

File: doc/src/preprocessor/contract_inheritance.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+pub const CONTRACT_INHERITANCE_ID: PreprocessorId = PreprocessorId("contract_inheritance");
+
+/// The contract inheritance preprocessor.
+/// It matches the documents with inner [`ParseSource::Contract`](crate::ParseSource) elements,
+/// iterates over their [Base](solang_parser::pt::Base)s and attempts
+/// to link them with the paths of the other contract documents.
+///
+/// This preprocessor writes to [Document]'s context.
+#[derive(Debug)]
+pub struct ContractInheritance;
```
Comment (mattsse) [inline doc/src/preprocessor/contract_inheritance.rs:15]:
```suggestion
#[derive(Default, Debug)]
#[non_exhaustive]
pub struct ContractInheritance;
```

File: doc/src/preprocessor/git_source.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
+use std::path::PathBuf;
+
```
Comment (mattsse) [inline doc/src/preprocessor/git_source.rs:2]:
```suggestion
```

**Review by mattsse — CHANGES_REQUESTED** [review]
only did a high level review, just because this is quite huge, but it's very readable and you can understand how this is supposed to work just by reading this.

smol nits, otherwise lgtm

**Comment by jameswenzel** [comment]
Running on the [Seaport 1.2 branch](https://github.com/ProjectOpenSea/seaport/tree/1.2) results in a rather unhelpful error:

```
❯ forge doc --serve
Error: 
Failed to parse source: unknown comment tag: 
```

**Comment by sambacha** [comment]
https://github.com/foundry-rs/foundry/blob/7c135cd969092710ad53d2532c3113a9a4517344/doc/src/writer/markdown.rs#L21:5

Only `FencedCodeBlock`, as there are two different kinds of Code Blocks: Fenced and Indentation based. I don't think you want to support Indentation-based code blocks for generated output. 

Can we add reference links for each natspec section back to the relevant source lineos?

**Review by gakonst — APPROVED** [review]
Absolutely incredible. Would prefer if we used Axum instead of Warp. Love the Preprocessor keeping things clean.

**Pull request merged**